packages/hardhat-core/test/internal/solidity/compiler/downloader.ts

```diff
@@ -316,6 +316,98 @@ describe("Compiler downloader", function () {
         });
       });
     });
+
+    describe("When downloading fails", function () {
+      describe("getCompilersList", function () {
+        it("retries on a failed download", async function () {
+          let downloadCount = 0;
+          const failingDownloader = (mockDownloader = new CompilerDownloader(
+            compilersDir,
+            {
+              download: async () => {
+                if (downloadCount === 1) {
+                  await saveMockCompilersList();
+                } else {
+                  await saveMalformedCompilersList();
+                }
+
+                downloadCount++;
+              },
+              forceSolcJs: true,
+            }
+          ));
+
+          await failingDownloader.getCompilersList(CompilerPlatform.WASM);
+          assert.equal(downloadCount, 2);
+        });
+
+        it("errors on too many failed downloads", async function () {
+          let downloadCount = 0;
+          const failingDownloader = new CompilerDownloader(compilersDir, {
+            download: async () => {
+              await saveMalformedCompilersList();
+
+              downloadCount++;
+            },
+            forceSolcJs: true,
+          });
+
+          try {
+            await failingDownloader.getCompilersList(CompilerPlatform.WASM);
+          } catch (error) {
+            assert.isTrue(error instanceof SyntaxError);
+          }
```

Comment:
This doesn't quite work because this doesn't check that there is actually an error. You can use `isRejected` for this, see for example:

https://github.com/nomiclabs/hardhat/blob/f911aa3b76a61835a1a83740065370fc6ab79183/packages/hardhat-core/test/internal/core/providers/accounts.ts#L412-L418

Reply:
Thanks for that. I couldn't find the async throws assertion on `chai`, I hadn't grasped we had `chai-as-promised`; and `isRejected` actually works rather than my homebrew.